src/runtime/uv/tcp.cpp

```diff
@@ -0,0 +1,587 @@
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_tcp_connect(b_obj_arg socket, b_obj_arg addr) {
+    lean_uv_tcp_socket_object * tcp_socket = lean_to_uv_tcp_socket(socket);
+
+    sockaddr addr_ptr;
+    lean_socket_addr_to_sockaddr(addr, &addr_ptr);
+
+    lean_object * promise = create_promise();
+
+    uv_connect_t * uv_connect = (uv_connect_t*)malloc(sizeof(uv_connect_t));
+
+    uv_connect->data = new tcp_connect_data { promise, socket };
+    lean_inc(promise);
+
+    event_loop_lock(&global_ev);
+
+    // The event loop owns the socket.
+    lean_inc(socket);
+
+    int result = uv_tcp_connect(uv_connect, tcp_socket->m_uv_tcp, (const struct sockaddr *)&addr_ptr, [](uv_connect_t* req, int status) {
+        tcp_connect_data* tup = (tcp_connect_data*) req->data;
+        resolve_promise_with_status(tup->promise, status);
+
+        // The event loop does not own the object anymore.
+        lean_dec(tup->socket);
+
+        free(req->data);
+        free(req);
+    });
+
+    event_loop_unlock(&global_ev);
+
+    if (result < 0) {
+        free(uv_connect);
+        free(uv_connect->data);
+
+        return lean_io_result_mk_error(lean_decode_uv_error(result, mk_string("failed to connect")));
+    }
+
+    return lean_io_result_mk_ok(promise);
+}
+
+/* Std.Internal.UV.Tcp.Socket.send (socket : @& Socket) (data : ByteArray) : IO (IO.Promise (Except IO.Error Unit)) */
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_tcp_send(b_obj_arg socket, obj_arg data) {
+    lean_uv_tcp_socket_object * tcp_socket = lean_to_uv_tcp_socket(socket);
+
+    size_t data_len = lean_sarray_size(data);
+    char * data_str = (char *)lean_sarray_cptr(data);
+
+    uv_buf_t buf = uv_buf_init(data_str, data_len);
+
+    lean_object * promise = create_promise();
+
+    uv_write_t * write_uv = (uv_write_t*)malloc(sizeof(uv_write_t));
+
+    write_uv->data = new tcp_send_data { promise, data, socket };
+    lean_inc(promise);
+
+    event_loop_lock(&global_ev);
+
+    // The event loop owns the socket.
+    lean_inc(socket);
+
+    int result = uv_write(write_uv, (uv_stream_t*)tcp_socket->m_uv_tcp, &buf, 1, [](uv_write_t * req, int status) {
+        tcp_send_data * tup = (tcp_send_data*) req->data;
+
+        resolve_promise_with_status(tup->promise, status);
+
+        // The event loop does not own the object anymore.
+        lean_dec(tup->socket);
+        lean_dec(tup->data);
+
+        free(req->data);
+        free(req);
+    });
+
+    event_loop_unlock(&global_ev);
+
+    if (result < 0) {
+        free(write_uv);
+        free(write_uv->data);
+
+        return lean_io_result_mk_error(lean_decode_uv_error(result, mk_string("failed to write")));
+    }
+
+    return lean_io_result_mk_ok(promise);
+}
+
+/* Std.Internal.UV.Tcp.Socket.recv (socket : @& Socket) : IO (IO.Promise (Except IO.Error ByteArray)) */
+extern "C" LEAN_EXPORT lean_obj_res lean_uv_tcp_recv(b_obj_arg socket) {
```

Comment:
As we seem to be approaching a final version of this now. It would be great if we could pass a `size : USize` into `recv` to tell the size of the buf to allocate instead of always taking libuvs suggestion. That should be doable right?